heft-plugins/heft-webpack5-plugin/src/WebpackConfigurationLoader.ts

```diff
@@ -16,58 +16,61 @@ interface IWebpackConfigFunctionEnv {
+  | TWebpack.Configuration[]
+  | Promise<TWebpack.Configuration>
+  | Promise<TWebpack.Configuration[]>
+  | ((env: IWebpackConfigFunctionEnv) => TWebpack.Configuration | TWebpack.Configuration[])
+  | ((env: IWebpackConfigFunctionEnv) => Promise<TWebpack.Configuration | TWebpack.Configuration[]>);
 type IWebpackConfigJs = IWebpackConfigJsExport | { default: IWebpackConfigJsExport };
 
 const WEBPACK_CONFIG_FILENAME: string = 'webpack.config.js';
 const WEBPACK_DEV_CONFIG_FILENAME: string = 'webpack.dev.config.js';
 
 export class WebpackConfigurationLoader {
-  public static async tryLoadWebpackConfigAsync(
-    logger: ScopedLogger,
-    buildFolder: string,
-    buildProperties: IBuildStageProperties
-  ): Promise<IWebpackConfiguration | undefined> {
+  private _logger: IScopedLogger;
+  private _production: boolean;
+  private _serveMode: boolean;
+
+  public constructor(logger: IScopedLogger, production: boolean, serveMode: boolean) {
+    this._logger = logger;
+    this._production = production;
+    this._serveMode = serveMode;
+  }
+
+  public async tryLoadWebpackConfigAsync(buildFolder: string): Promise<IWebpackConfiguration | undefined> {
     // TODO: Eventually replace this custom logic with a call to this utility in in webpack-cli:
     // https://github.com/webpack/webpack-cli/blob/next/packages/webpack-cli/lib/groups/ConfigGroup.js
 
     let webpackConfigJs: IWebpackConfigJs | undefined;
 
     try {
-      if (buildProperties.serveMode) {
-        logger.terminal.writeVerboseLine(
+      if (this._serveMode) {
+        this._logger.terminal.writeVerboseLine(
           `Attempting to load webpack configuration from "${WEBPACK_DEV_CONFIG_FILENAME}".`
         );
-        webpackConfigJs = WebpackConfigurationLoader._tryLoadWebpackConfiguration(
+        webpackConfigJs = await this._tryLoadWebpackConfigurationAsync(
           buildFolder,
           WEBPACK_DEV_CONFIG_FILENAME
         );
       }
 
       if (!webpackConfigJs) {
-        logger.terminal.writeVerboseLine(
+        this._logger.terminal.writeVerboseLine(
           `Attempting to load webpack configuration from "${WEBPACK_CONFIG_FILENAME}".`
         );
-        webpackConfigJs = WebpackConfigurationLoader._tryLoadWebpackConfiguration(
-          buildFolder,
-          WEBPACK_CONFIG_FILENAME
-        );
+        webpackConfigJs = await this._tryLoadWebpackConfigurationAsync(buildFolder, WEBPACK_CONFIG_FILENAME);
       }
     } catch (error) {
-      logger.emitError(error as Error);
+      this._logger.emitError(error as Error);
     }
 
     if (webpackConfigJs) {
       const webpackConfig: IWebpackConfigJsExport =
         (webpackConfigJs as { default: IWebpackConfigJsExport }).default || webpackConfigJs;
 
       if (typeof webpackConfig === 'function') {
-        return webpackConfig({ prod: buildProperties.production, production: buildProperties.production });
+        return webpackConfig({ prod: this._production, production: this._production });
```

Comment:
Same comment as the webpack 4 plugin, would be nice if the webpack.config.js file had access to the Heft task environment

Reply:
Is there a need for the `taskSession`/`heftConfiguration`/`pluginOptions` in addition to the `webpack` package?

Reply:
So that package-level webpack configuration can reference information from the environment, such as the list of requested locales, the containing package name, the project folder, etc.